runtime/common/logging_xwalk.cc

```diff
@@ -0,0 +1,205 @@
+// InitXwalkLogging() and the beginning of CleanupXwalkLogging().
+bool xwalk_logging_initialized_ = false;
+
+// Set if we called InitXwalkLogging() but failed to initialize.
+bool xwalk_logging_failed_ = false;
+
+// This should be true for exactly the period between the end of
+// InitXwalkLogging() and the beginning of CleanupXwalkLogging().
+bool xwalk_logging_redirected_ = false;
+
+#if defined(OS_WIN)
+// {7FE69228-633E-4f06-80C1-527FEA23E3A7}
+const GUID kXwalkTraceProviderName = {
+    0x7fe69228, 0x633e, 0x4f06,
+        { 0x80, 0xc1, 0x52, 0x7f, 0xea, 0x23, 0xe3, 0xa7 } };
+#endif
+
+}  // anonymous namespace
+
+namespace logging {
+
+LoggingDestination DetermineLogMode(const base::CommandLine& command_line) {
+  // only use OutputDebugString in debug mode
+#ifdef NDEBUG
+  bool enable_logging = false;
+  const char *kInvertLoggingSwitch = switches::kEnableLogging;
+  const logging::LoggingDestination kDefaultLoggingMode = logging::LOG_TO_FILE;
+#else
+  bool enable_logging = true;
+  const char *kInvertLoggingSwitch = switches::kDisableLogging;
+  const logging::LoggingDestination kDefaultLoggingMode = logging::LOG_TO_ALL;
+#endif
+
+  if (command_line.HasSwitch(kInvertLoggingSwitch))
+    enable_logging = !enable_logging;
+
+  logging::LoggingDestination log_mode;
+  if (enable_logging) {
+    // Let --enable-logging=stderr force only stderr, particularly useful for
+    // non-debug builds where otherwise you can't get logs to stderr at all.
+    if (command_line.GetSwitchValueASCII(switches::kEnableLogging) == "stderr")
+      log_mode = logging::LOG_TO_SYSTEM_DEBUG_LOG;
+    else
+      log_mode = kDefaultLoggingMode;
+  } else {
+    log_mode = logging::LOG_NONE;
+  }
+  return log_mode;
+}
+
+void InitXwalkLogging(const base::CommandLine& command_line,
+                      OldFileDeletionState delete_old_log_file) {
+  DCHECK(!xwalk_logging_initialized_) <<
+      "Attempted to initialize logging when it was already initialized.";
+
+  LoggingDestination logging_dest = DetermineLogMode(command_line);
+  LogLockingState log_locking_state = LOCK_LOG_FILE;
+  base::FilePath log_path;
+
+  // Don't resolve the log path unless we need to. Otherwise we leave an open
+  // ALPC handle after sandbox lockdown on Windows.
+  if ((logging_dest & LOG_TO_FILE) != 0) {
+    log_path = GetLogFileName();
+  } else {
+    log_locking_state = DONT_LOCK_LOG_FILE;
+  }
+
+  logging::LoggingSettings settings;
+  settings.logging_dest = logging_dest;
+  settings.log_file = log_path.value().c_str();
+  settings.lock_log = log_locking_state;
+  settings.delete_old = delete_old_log_file;
+  bool success = logging::InitLogging(settings);
+
+  if (!success) {
+    DPLOG(ERROR) << "Unable to initialize logging to " << log_path.value();
+    xwalk_logging_failed_ = true;
+    return;
+  }
+
+  // Default to showing error dialogs.
+  if (!base::CommandLine::ForCurrentProcess()->HasSwitch(
+      switches::kNoErrorDialogs))
```

Comment:
My `GYP_DEFINES` are `component=shared_library disable_nacl=1`. Building the `xwalk_builder` in Release mode fails here:

```
[20/24] LINK xwalk
FAILED: ../../third_party/llvm-build/Release+Asserts/bin/clang++ -Wl,-z,now -Wl,-z,relro -Wl,--fatal-warnings -Wl,-z,defs -pthread -Wl,-z,noexecstack -fPIC -fuse-ld=gold -B/data/src/crosswalk/crosswalk/src/third_party/binutils/Linux_x64/Release/bin -Wl,--disable-new-dtags -L. -Wl,-uIsHeapProfilerRunning,-uProfilerStart -Wl,-u_Z21InitialMallocHook_NewPKvj,-u_Z22InitialMallocHook_MMapPKvS0_jiiix,-u_Z22InitialMallocHook_SbrkPKvi -Wl,-u_Z21InitialMallocHook_NewPKvm,-u_Z22InitialMallocHook_MMapPKvS0_miiil,-u_Z22InitialMallocHook_SbrkPKvl -Wl,-u_ZN15HeapLeakChecker12IgnoreObjectEPKv,-u_ZN15HeapLeakChecker14UnIgnoreObjectEPKv -m64 -Wl,--detect-odr-violations -Wl,--icf=all -Wl,-O1 -Wl,--as-needed -Wl,--gc-sections -Wl,-rpath=\$ORIGIN/lib/ -Wl,-rpath-link=lib/ -o xwalk -Wl,--start-group obj/xwalk/runtime/app/xwalk.xwalk_main.o obj/xwalk/libxwalk_runtime.a obj/base/third_party/dynamic_annotations/libdynamic_annotations.a obj/components/libautofill_content_browser.a obj/google_apis/libgoogle_apis.a obj/third_party/libxml/libxml2.a obj/third_party/zlib/libchrome_zlib.a obj/third_party/zlib/libzlib_x86_simd.a obj/third_party/libphonenumber/libphonenumber.a obj/third_party/libphonenumber/libphonenumber_without_metadata.a obj/components/libautofill_content_common.a obj/components/libautofill_core_common.a obj/components/libautofill_content_risk_proto.a obj/components/libautofill_core_browser.a obj/third_party/fips181/libfips181.a obj/third_party/libjingle/libjingle.a obj/third_party/webrtc/base/librtc_base.a obj/third_party/webrtc/libwebrtc_common.a obj/third_party/webrtc/base/librtc_base_approved.a obj/third_party/webrtc/libjingle/xmllite/librtc_xmllite.a obj/third_party/libjingle/libjingle_p2p_constants.a obj/third_party/libaddressinput/libaddressinput_util.a obj/third_party/re2/libre2.a obj/components/libinfobars_core.a obj/components/libos_crypt.a obj/components/libpref_registry.a obj/components/librappor.a obj/third_party/smhasher/libcityhash.a obj/components/libmetrics.a obj/components/libcomponent_metrics_proto.a obj/components/libvariations.a obj/third_party/mt19937ar/libmt19937ar.a obj/components/libsignin_core_browser.a obj/third_party/cacheinvalidation/libcacheinvalidation_proto_cpp.a obj/components/libcontent_settings_core_browser.a obj/components/libcontent_settings_core_common.a obj/components/libplugins_common.a obj/components/url_formatter/liburl_formatter.a obj/components/libgoogle_core_browser.a obj/components/libinvalidation_public.a obj/third_party/cacheinvalidation/libcacheinvalidation.a obj/components/libsignin_core_common.a obj/components/libvariations_http_provider.a obj/components/libautofill_content_renderer.a obj/components/libcdm_renderer.a obj/components/libcdm_common.a obj/components/libdevtools_http_handler.a obj/components/libdevtools_discovery.a obj/net/libhttp_server.a obj/components/libvisitedlink_browser.a obj/components/libvisitedlink_common.a obj/components/libvisitedlink_renderer.a obj/xwalk/libxwalk_application_lib.a obj/third_party/zlib/google/libzip.a obj/third_party/zlib/libminizip.a obj/xwalk/application/common/libxwalk_application_common_lib.a obj/xwalk/extensions/libxwalk_extensions.a obj/xwalk/sysapps/libsysapps.a obj/tools/json_schema_compiler/libapi_gen_util.a obj/components/libstorage_monitor.a obj/device/media_transfer_protocol/libdevice_media_transfer_protocol.a obj/device/media_transfer_protocol/libmtp_file_entry_proto.a obj/device/media_transfer_protocol/libmtp_storage_info_proto.a obj/device/udev_linux/libudev_linux.a obj/build/linux/libudev.a obj/base/allocator/liballocator.a obj/components/libapp_modal.a obj/components/libconstrained_window.a obj/components/libweb_modal.a obj/components/libguest_view_browser.a obj/components/libguest_view_common.a obj/components/libui_zoom.a obj/components/libpdf_renderer.a obj/components/libpdf_common.a -Wl,--end-group lib/libui_base.so lib/libstorage_common.so lib/libboringssl.so lib/libgfx_ipc.so lib/libwebview.so lib/libprotobuf_lite.so lib/libkeyed_service_core.so lib/libui_base_ime.so lib/libgl_wrapper.so lib/libcc.so lib/libffmpeg.so lib/libbase_prefs.so lib/libskia.so lib/libppapi_proxy.so lib/liburl_lib.so lib/libwm.so lib/libbase.so lib/libgfx_geometry.so lib/libnet.so lib/libcrcrypto.so lib/libppapi_host.so lib/libgfx.so lib/libaura.so lib/libv8.so lib/libgin.so lib/libicui18n.so lib/libwebdata_common.so lib/libipc.so lib/libsync_core.so lib/libblink_common.so lib/libicuuc.so lib/libsync_proto.so lib/libevents.so lib/libviews.so lib/libdbus.so lib/libblink_platform.so lib/libuser_prefs.so lib/libsql.so lib/libblink_web.so lib/libgtk2ui.so lib/libsnapshot.so lib/libstorage.so lib/libmedia.so lib/libppapi_shared.so lib/libbase_i18n.so lib/libcontent.so lib/libshell_dialogs.so -lrt -ldl -lm -lexpat -lXext -lX11 -lXcomposite -lXrender -lasound -ldbus-1 -lz -lnotify -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lcups -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err -lpthread -lcrypt
obj/xwalk/libxwalk_runtime.a(obj/xwalk/runtime/common/xwalk_runtime.logging_xwalk.o):../../xwalk/runtime/common/logging_xwalk.cc:function logging::InitXwalkLogging(base::CommandLine const&, logging::OldFileDeletionState): error: undefined reference to 'switches::kNoErrorDialogs'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```